### ISSUE TITLE
Add support for fetching pages from browser cache

### ISSUE THREAD

**Issue by sidney** [JimmXinu/FanFicFare#635]
I'm creating this issue to branch the discussion off from issue #622 

So far I have an implementation using the Chrome Simple Cache format which I have tested to work on macOS Catalina and on Linux Ubuntu 20.04 with the latest version of Chrome. I also installed Chrome on Windows 10 and verified that it uses a different cache format. The Simple Cache format can be distinguished by its use of file names that are 16 hex digits followed by _0, a file named index that is 24 bytes in size, and the actual larger index file in index-dir/the-real-index. On Windows 10 I see the Cache directory has files named, data_n,  f_nnnnnn where nnnnnn are sequential hex numbers, and the index being in a file named index, no index-dir directory.

When this is sufficiently done, I'll upload it as a PR to [ffnet-chrome-cache-fetch](https://github.com/JimmXinu/FanFicFare/tree/ffnet-chrome-cache-fetch)

**Comment by JimmXinu** [comment]
FYI: Commit 8b1da6f6ec0a17db8a8655eba2b692f97566ca4b changes adapter_fanfictionnet to keep story title in storyUrl and use urltitle in chapter fetchs, but not when stored in epub.

**Comment by JimmXinu** [comment]
I've rebased ffnet-chrome-cache-fetch onto master, but I haven't tested it.

**Comment by sidney** [comment]
I delayed uploading a PR because I decided to write it as a general class that uses my Simple Cache code and the Chromagnon code to determine what format the cache directory contains and choose to use the right one. That way adding Firefox cache support in the future is just a matter of writing a subclass that implements a couple of methods to read that format. I'll rebase as you did and keep going with it.

Do I need to be concerned about maintaining Python 2 compatibility, or is it ok now to write Python 3 only code? I'm an experienced programmer but never did serious work in Python, so I'm not familiar with the idioms of code that works in both 2 and 3.

**Comment by JimmXinu** [comment]
That sounds like a good idea.  Since we're not time crunched anymore, I'll probably redo the FFF parts to move out of the ffnet adapter and into the more general code before it ever gets released.  Any adapter that does GETs could benefit, but not, I expect, POSTs.

In theory, Python2 isn't supported anymore by the python community.  And I am starting to find libraries that aren't py2 compatible anymore.

In practice, Calibre only switched to py3 with version 5 in Sept 2020 and [nearly half](https://calibre-ebook.com/dynamic/calibre-usage) (49% when I write this) of Calibre users are still on older, py2 versions.  

I do still consider v4 support to be critical at this time for main-line FFF.  

I have managed to keep FFF running back to Calibre v2.85.1, which should cover ~95% of Calibre users.  So far, keeping it running on v2 & v3 hasn't been significantly worse than v4.

OTOH, this isn't necessarily something that needs to be in the main-line, since it will be a work-around and/or performance enhancement, depending how it works out.  It was really kind of a combination clever-idea / straw-grasp when I started on it.

**Comment by sidney** [comment]
Ok, I'll ignore Python 2 considerations for now, then after I have enough to upload a PR for this branch, I'll let my IDE (IntelliJ IDEA) give me the hints it nags me about Python 2 incompatibilities, and do another pass for that.

BTW, commit 39d23c8c9865d7fb223631bf3c949fa57bb9cb8b somehow not only commented out the `check-next-chapter` code, but also reverted it in the comment to what it was before you added `self.urltitle` to the URL. So when you thought you were putting the code back in by uncommenting it in commit ea4cf245ac1f4773846abf649697d44227994b24 you ended up removing the `urltitle` changes from it.

It looks like the only thing you need to fix is in `adapter_fanfictionnet.py` in the `ffnet-chrome-cache-fetch` branch, from

```
                tryurl = "https://%s/s/%s/%d/"%(self.getSiteDomain(),
                                                self.story.getMetadata('storyId'),
                                                chapcount+1)

```
to

```
                tryurl = "https://%s/s/%s/%d/%s"%(self.getSiteDomain(),
                                                  self.story.getMetadata('storyId'),
                                                  chapcount+1,
                                                  self.urltitle)

```

**Comment by sidney** [comment]
@JimmXinu Also, `fanficfare/chromagnon/cacheParse.py` in the branch ended up with merge conflict sections after the rebase, mostly because of the change in import of brotli and the profiling code. You can see a bunch of `<<<<<<< HEAD` sections in it.

**Comment by JimmXinu** [comment]
I tried to rebase it and had a half dozen conflicts that each came up sequentially--I probably should have merged instead.  I don't work with other people (or even branches) very often.

I see I also checked in conflicts in the chromagnon code.  Oops.

**Comment by JimmXinu** [comment]
Okay, fixed.  I think.  It worked in CLI, anyway.
9572c25c0bf9ea0fc2e21247f4e33304bd998b67

**Comment by sidney** [comment]
I am confused. Even when I try the code with none of my changes using Chrome on Windows 10, I get URL not found in cache for all the URLs. I added logging to the code and I see that the keys in `self.hash_cache` all look like this

`'_dk_https://fanfiction.net https://fanfiction.net https://www.fanfiction.net/s/12213213/1/The-Need-to-Become-Stronger'`

I tracked it down to where the dictionary is created and saw that keys came from reading file `data_1` and looking inside that file I see that all the keys are in the format

`_dk_https://example.com https://example.com https://www.example.com/rest/of/url`

My WIndows 10 VM is running Chrome version 88, but when I first tested it was version 87. I reverted to a VM snapshot that still had Chrome version 87 and the cache looked the same. What version are you testing with, and what do you see in the `data_1` file? I think that changing the regexp used to get `normkey` in the class `ChromeCache` `__init__` in `cacheParse.py` could work around it if it matched on just the end URL part, but I didn't want to touch that without understanding how it worked for you as is.

**Comment by sidney** [comment]
It works if I change

`            normkey = re.sub(r'^(https://www.fanfiction.net/s/\d+/\d+/).+$',r'\1',key)`

to

`            normkey = re.sub(r'^.*(https://www.fanfiction.net/s/\d+/\d+/[^/]+).*$',r'\1',key)`

But I don't think that is the right place. Might as well fix up the key when it is first created in CacheEntry._init_ by extracting the URL from the end of self.key before returning. If you do that I don't think there will really be any need for normkey at all.

I'll try that.

**Comment by sidney** [comment]
I submitted PR #638 
Tested on Windows and macOS formats, in GUI and CLI

**Comment by JimmXinu** [comment]
FYI, I started working on making ffnet-chrome-cache-fetch work in older, python2 Calibre versions but stopped when I realized `brotlipython` is also py3 only.  Have you looked in there?  It's a bunch of bit and byte manipulation code that apparently came from a nearly one-to-one translation from C code.  That's a level of pain I'm not interested in today.

(For anyone reading this in the far future, PR #638 was merged into  [ffnet-chrome-cache-fetch](https://github.com/JimmXinu/FanFicFare/tree/ffnet-chrome-cache-fetch).)

**Comment by sidney** [comment]
To continue from remarks you made in PR #642 it probably will be necessary to access the entire cache when initializing it in order to build an index that has the actual keys. In SimpleCache, the index file `index-dir/the-real-index` only contains a list of hashes without the keys that generated them. It is of some use for the browser to keep an index file with just hashes and timestamps, but not for our purposes. In the disk cache format that Chromagnon uses, It is also necessary to go through the entire cache. The index file only contains the hashes and keys of first buckets. The only way to find entries that are in the subsequent buckets of a collision list is by following the pointers that are in the cache entry files. And they all have to be read because the only indication that there are no further buckets is that the cache entry for the first one does not contain a link.

I can see how you would want to initialize the browser cache instance just once when starting a job that may use multiple adapters. If you do that, then it would work to extract the real URL from the key string in the cache using `key_string.split(' ')[-1]` which will work whether the key string is the old format of the actual URL or the new one with `_dk_`.

What is tricky in any case is matching up the URL being fetched with the URL used in the cache key. That would only be possible if the site really does let you reliably test if a URL is valid for a chapter of the story being fetched and tranfsorm it to a normalized form. Are there any adapters that for which the site has no discernable pattern in the chapter links so you don't know what you will need until you read the chapter link? Actually, that's not so bad... How about this? Initialize a dictionary mapping a set of URL strings from the keys found in all the cache entry files (only have to read headers from each file, not the entire file) to the actual keys used in the cache entries. When initializing an adapter, extract from that just the URLs that are suitable for the adapter, using whichever of the functions you have that is the best test of that. If the adapter is able to have a good normalize URL function, then use that to make a dictionary mapping the normalized URL to the key in the cache. If there is not a good normalize URL function, then you would need to search through the set of URLs that were extracted to find one that is equivalent to the URL you want to fetch. Is that something that would make sense to do?

**Comment by sidney** [comment]
I'm working on a Python2 compatible Python only brotli library that I am translating by hand from the `brotli.js` library on Github, which is itself a hand translation of the C reference  implementation. It seems to be working, passing all the tests that are in the reference implementation. I want to do a bit of profiling before I commit it to a public github repo.

Here are the current test results that also include time comparison with `brotli` and `brotlipython`. As you can see it is still way slower than `brotli`, which uses thin python bindings to C, but faster than `brotlipython`. Interestingly enough, it is noticeably faster in Python 2 than in Python 3, whereas brotli is faster in 3 than in 2. Ignore the difference in total time between the two, because the Python 2 tests do not include running brotlipython.  

In Python 3.9
```
BrotliPythonDecompresser % PYTHONPATH=. python test/decompress_test.py
File '10x10y' Times msec   C ref: 0.059, brotlipy: 181.769, this test: 134.286
.File '64x' Times msec   C ref: 0.032, brotlipy: 195.136, this test: 119.064
.File 'alice29.txt' Times msec   C ref: 0.774, brotlipy: 1208.458, this test: 389.195
.File 'asyoulik.txt' Times msec   C ref: 0.739, brotlipy: 1106.063, this test: 400.097
.File 'backward65536' Times msec   C ref: 1.824, brotlipy: 303.111, this test: 168.845
.File 'compressed_file' Times msec   C ref: 0.066, brotlipy: 286.998, this test: 143.776
.File 'compressed_repeated' Times msec   C ref: 0.306, brotlipy: 811.037, this test: 308.298
.File 'empty' Times msec   C ref: 0.020, brotlipy: 197.771, this test: 133.209
.File 'lcet10.txt' Times msec   C ref: 1.658, brotlipy: 2753.210, this test: 825.013
.File 'mapsdatazrh' Times msec   C ref: 1.970, brotlipy: 2313.422, this test: 755.326
.File 'monkey' Times msec   C ref: 0.039, brotlipy: 209.675, this test: 137.985
.File 'plrabn12.txt' Times msec   C ref: 2.205, brotlipy: 5215.020, this test: 1171.739
.File 'quickfox' Times msec   C ref: 0.027, brotlipy: 195.307, this test: 150.799
.File 'quickfox_repeated' Times msec   C ref: 1.647, brotlipy: 530.001, this test: 222.075
.File 'random_org_10k.bin' Times msec   C ref: 0.039, brotlipy: 248.643, this test: 137.719
.File 'ukkonooa' Times msec   C ref: 0.042, brotlipy: 262.079, this test: 145.323
.File 'x' Times msec   C ref: 0.019, brotlipy: 222.192, this test: 143.891
.File 'xyzzy' Times msec   C ref: 0.022, brotlipy: 214.038, this test: 94.930
.File 'zeros' Times msec   C ref: 10.430, brotlipy: 619.925, this test: 255.128
.......................................
----------------------------------------------------------------------
Ran 57 tests in 37.767s
```

In Python 2.7

```
BrotliPythonDecompresser % PYTHONPATH=. python test/decompress_test.py
File '10x10y' Times msec   C ref: 0.071, brotlipy: nan, this test: 96.354
.File '64x' Times msec   C ref: 0.044, brotlipy: nan, this test: 82.550
.File 'alice29.txt' Times msec   C ref: 0.808, brotlipy: nan, this test: 258.398
.File 'asyoulik.txt' Times msec   C ref: 0.852, brotlipy: nan, this test: 260.489
.File 'backward65536' Times msec   C ref: 1.532, brotlipy: nan, this test: 110.963
.File 'compressed_file' Times msec   C ref: 0.057, brotlipy: nan, this test: 80.572
.File 'compressed_repeated' Times msec   C ref: 0.371, brotlipy: nan, this test: 214.040
.File 'empty' Times msec   C ref: 0.021, brotlipy: nan, this test: 81.011
.File 'lcet10.txt' Times msec   C ref: 1.904, brotlipy: nan, this test: 540.864
.File 'mapsdatazrh' Times msec   C ref: 2.330, brotlipy: nan, this test: 512.868
.File 'monkey' Times msec   C ref: 0.044, brotlipy: nan, this test: 84.830
.File 'plrabn12.txt' Times msec   C ref: 2.319, brotlipy: nan, this test: 656.162
.File 'quickfox' Times msec   C ref: 0.030, brotlipy: nan, this test: 79.159
.File 'quickfox_repeated' Times msec   C ref: 1.463, brotlipy: nan, this test: 121.078
.File 'random_org_10k.bin' Times msec   C ref: 0.035, brotlipy: nan, this test: 82.070
.File 'ukkonooa' Times msec   C ref: 0.042, brotlipy: nan, this test: 79.417
.File 'x' Times msec   C ref: 0.030, brotlipy: nan, this test: 78.173
.File 'xyzzy' Times msec   C ref: 0.027, brotlipy: nan, this test: 78.555
.File 'zeros' Times msec   C ref: 9.193, brotlipy: nan, this test: 133.766
.......................................
----------------------------------------------------------------------
Ran 57 tests in 14.432s
```

**Comment by sidney** [comment]
It runs well enough for now, I'll get to profiling later. See https://github.com/sidney/brotlidecpy

**Comment by JimmXinu** [comment]
This is great!  Many thanks for doing this; I've already been able to test and use it in both py2 and py3 Calibre.  

I'm most of the way through refactor/rewriting the entire http fetching code in FFF and then I'll merge the chrome cache fetching in properly.

I'll push an updated ffnet-chrome-cache-fetch branch soon with py2/py3 compatibility.  But I have a suggestion first about packaging brotlidecpy.  [Pull request](https://github.com/sidney/brotlidecpy/pull/1).

**Comment by JimmXinu** [comment]
Updated https://github.com/JimmXinu/FanFicFare/tree/ffnet-chrome-cache-fetch

I already including my package rename PR.  If you decide against it, I'll change the FFF copy back to match.

**Comment by JimmXinu** [comment]
Okay, it took all week, but I bludgeoned my ancient code into a somewhat more pleasing shape ([master_fetch_refactor](https://github.com/JimmXinu/FanFicFare/tree/master_fetch_refactor)).

After that, I merged in the chrome cache code and was able to slot it in relatively quickly in branch: [master_fetch_refactor_merge_chrome_cache](https://github.com/JimmXinu/FanFicFare/tree/master_fetch_refactor_merge_chrome_cache)

Some references:
- `configurable.py`:977, where BrowserCache is currently created.
- `fetcher.py`:258, the decorator that applies BrowserCache to the `do_request` chain

I have *not* tried to bring in any changes from PR #642 yet.

Things that still need attention:
- What from PR #642 should be added?
- The `_dk_` key look up is _slightly_ better.  A little.
- Does browser cache need to pre-read all the entries in the cache to index with 'normalized' URLs?  And the requested URLs?  And how is 'normal' determined?
- If so, BrowserCache needs to be able to be shared between Configuration/Fetchers.  Singleton under the covers?  Separate the 'this is where the browser cache is' and 'use the browser cache' settings.  Search for 'pagecache' in `cli.py` and `fff_plugin.py` and `jobs.py` to see how the basic page cache is shared.
- I haven't carried over the network reduction changes for adapter_fanfictionnet; no cover, not fetching crossover names.
- Why is my Chrome using that ancient cache format?  It would be nice to only have one.  Clearing cache and making a new profile still give me the old one.

This list is more for me--I'm not expecting you do them all. :-) But I did want to share that I have been working towards it.

**Comment by sidney** [comment]
My Mac was running Chrome version 88, had simple cache format, and did not have the new partitioning cache feature (the _dk_ thing). There was an upgrade available for Chrome, which was still version 88 but a newer branch/build. After the upgrade it has the partitioning cache feature.  On my Windows 10 VM, I had Chrome 87, and it had the older format disk cache which I've found is called the blockfile cache backend, but it did have partitioning cache. I let Chrome upgrade to version 88, verified that it is the same sub-version as the update on the mac - `chrome://version` shows `88.0.4324.104` (the 104 part doesn't mean much. I think 4324 is the release branch that you can compare cross-platform) and it still had blockfile cache format even after deleting all profiles and letting them get created again. I even tried uninstalling Chrome and removing all files, then re-installing with the latest version. The cache still got created in blockfile cache format.

I found this page that explains how to verify what type of cache Chrome is using and some other information about it.
https://developers.google.com/web/updates/2020/10/http-cache-partitioning
See the first FAQ on that page for how to capture a log and view it. They say how to see if the flag is set for cache partitioning. In addition, not mentioned there, the viewer they describe has a menu item on the left named Cache which displays the type of cache backend being used. My Mac demonstrated that the partitioning feature can be on or off independently of whether the backend is simple cache or blockfile cache.

As far as I can tell with both cache formats there is no getting around scanning the headers of all cache entries first, so you will need a way to do it once in a way that is shared by all processes of a job.

**Comment by sidney** [comment]
I pushed a PR to my brotlidecpy repo, merged it to master, then bumped the version number in master. The PR description has the summary of what I changed,

https://github.com/sidney/brotlidecpy/pull/2

**Comment by JimmXinu** [comment]
Got it, thanks.

**Comment by JimmXinu** [comment]
> As far as I can tell with both cache formats there is no getting around scanning the headers of all cache entries first, so you will need a way to do it once in a way that is shared by all processes of a job.

I agree; partitioned cache requires it.

I had hoped to only support Simple Cache and drop Blockfile cache.  But both your research and mine shows Blockfile cache in current use.  Users _were_ able to enable Simple Cache with `chrome://flags/#enable-simple-cache-backend`, but that flag has since been removed.  So we're apparently stuck with supporting both.

In addition to [Simple Cache](https://www.chromium.org/developers/design-documents/network-stack/disk-cache/very-simple-backend), there's a document on chromium.org about a [Disk cache 3.0](https://www.chromium.org/developers/design-documents/network-stack/disk-cache/disk-cache-v3), while the comments in the chromagnon only refer to [Blockfile cache](https://www.chromium.org/developers/design-documents/network-stack/disk-cache).

I can't find any evidence that this v3 was ever implemented.  The v3 doc says:
> Functionality specific to the new version is implemented by files located under src/net/disk_cache/v3.

But there's no such directory in [the chromium source](https://chromium.googlesource.com/chromium/src/+/refs/heads/master/net/disk_cache/).  The v3 document isn't dated that I could find, but there was a reference to the Disk cache 3.0 document in [a page from 2014](https://www.slideshare.net/cwdoh/cache-in-chromium-disk-cache).

So I'm going to ignore this disk cache v3 until it becomes a problem.

Thanks again for your help.

**Comment by sidney** [comment]
I released a new version of `brotlidecpy` https://github.com/sidney/brotlidecpy/releases/tag/v1.0.2
This tweaks/streamlines the most called inner loop function for a modest performance increase in Python 3. The improvement is much better when running under Python 2, but I don't see any way around the difference in speed processing small integers in Python 2 vs the fact that Python 3 uses unlimited length integer arithmetic even for small integers and there are no bit operations available on bytes without converting them to integers.

I also added a CHANGELOG.md

**Comment by JimmXinu** [comment]
Merged into master as of a78f8e94ee1b498cd8e0376ae9d6e1549e16ca60.

But I'll leave this issue open for the time being in case there's more to discuss.

**Comment by JimmXinu** [comment]
FYI, I've added an implementation now for Firefox cache2.

I'm going to go ahead and close this issue.  Many thanks to @sidney for all the invaluable help.
